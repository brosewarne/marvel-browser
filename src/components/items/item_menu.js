import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Typography from 'material-ui/Typography';

import { startsWithApiCall } from '../../api/api';
import { ItemTile } from './item_tile';
import { ItemGrid } from './item_grid';
import { PaginationMenu } from '../menus/pagination_menu';
import { AlphabetMenu } from '../menus/alphabet_menu';

import './items.css';
import '../../App.css';

import loading from '../../loading.svg';

class ItemMenu extends Component {
    static propTypes = {
        params: PropTypes.object.isRequired,
        itemType: PropTypes.string.isRequired
    };

    constructor(props) {
        super(props);

        this.state = {
            loading: true,
            startsWith: 'A',
            items: [],
            itemsPerPage: 9,
            currentPage: 1,
            totalPages: 0
        };
    }

    /**
     * Get the data for the set of items based on `startsWith`, `currentPage` and `itemsPerPage` and set
     * the item list and total items to internal state
     */
    getItemData = () => {
        const { startsWith, itemsPerPage, currentPage } = this.state;
        const offset = itemsPerPage * (currentPage - 1);
        // @todo handle AJAX request errors here
        startsWithApiCall(this.props.itemType, startsWith, itemsPerPage, offset).then((response) => {
            this.setState({
                loading: false,
                items: response.data.results,
                totalPages: Math.ceil(response.data.total / this.state.itemsPerPage)
            });
        });
    };

    /**
     * Get the initial set of item data when the page mounts.
     */
    componentWillMount = () => {
        this.setState({ ...this.props.params }, this.getItemData);
    };

    /**
     * Update the set of item data when the route changes
     * @param {Object} nextProps - The incoming set of props
     */
    componentWillReceiveProps = (nextProps) => {
        this.setState({ ...nextProps.params, loading: true }, this.getItemData);
    };

    /**
     * Get the pagination menu for the current set of items
     * @returns {Node} - The rendered PaginationMenu
     */
    getPaginationMenu = () => {
        const { totalPages, startsWith } = this.state;
        return (
            <PaginationMenu
                totalPages={totalPages}
                startsWith={startsWith}
                loading={this.state.loading}
            />
        );
    };

    /**
     * Get the pagination details for display - eg Page 1 of 10
     * @returns {Node} - The pagination details
     */
    getPaginationDetails = () => {
        return (
            !this.state.loading && (
                <Typography variant="headline">
                    <strong>Browsing {this.state.startsWith} {this.props.itemType} - Page {this.state.currentPage} of {this.state.totalPages}</strong>
                </Typography>
            )
        );
    };

    /**
     * Return an alphabet menu for viewing items that start with each letter of the alphabet
     * @returns {Node} - The rendered AlphabetMenu
     */
    getAlphabetMenu = () => {
        return <AlphabetMenu itemType={this.props.itemType} />;
    };

    /**
     * Create a set of item tiles for the current set of items.
     * @returns {Array} - The array of rendered ItemTile
     */
    getItemTiles = () => {
        return this.state.items.map((item) => {
            return <ItemTile item={item} itemType={this.props.itemType} key={`item_tile_${item.id}`} />;
        });
    };

    /**
     * Create a loading Image with text.
     * @returns {Node} - The loading image and text
     */
    getLoadingImage = () => {
        return (
            <div>
                <h1>
                    LOADING {this.props.itemType.toUpperCase()}.....
                </h1>
                <div>
                    <img src={loading} className="loadingSpinner" alt="loading" />
                </div>
            </div>
        );
    };

    /**
     * Create a grid of item tiles for the current set of items.
     * @returns {Node} - The item grid
     */
    getItemGrid = () => {
        if (this.state.loading) {
            return this.getLoadingImage();
        }
        const itemTiles = this.getItemTiles();
        return <ItemGrid itemTiles={itemTiles} />;
    };

    render() {
        return (
            <div>
                <section>
                    {this.getAlphabetMenu()}
                </section>
                <content>
                    {this.getPaginationDetails()}
                    {this.getItemGrid()}
                    {this.getPaginationMenu()}
                </content>
            </div>
        );
    }
}

export { ItemMenu };
