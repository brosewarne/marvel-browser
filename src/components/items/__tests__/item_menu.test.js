import React from 'react';
import { shallow, configure } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';

import { ItemMenu } from '../item_menu';

configure({ adapter: new Adapter() });

const createMockCharacters = (numCharacters) => {
    const mocks = [];
    for (let i = 0; i < numCharacters; i++) {
        mocks.push({
            id: i,
            name: `Character_${i}`,
            thumbnail: {
                path: '/a/href/to/an/image',
                extension: 'png'
            }
        });
    }
    return mocks;
};


describe('<ItemMenu />', () => {
    // no need to use real ItemTitle components here.
    const mockItemTiles = [
        <span key="c1">Character 1</span>,
        <span key="c2">Character 2</span>,
        <span key="c3">Character 3</span>,
        <span key="c4">Character 4</span>,
        <span key="c5">Character 5</span>,
        <span key="c6">Character 6</span>,
        <span key="c7">Character 7</span>,
        <span key="c8">Character 8</span>,
        <span key="c9">Character 9</span>,
        <span key="c10">Character 10</span>
    ];

    const params = {
        currentPage: 1,
        startsWith: 'A'
    };

    it('renders the loading spinner correctly', () => {
        const wrapper = shallow(
            <ItemMenu params={params} itemType="characters" />
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('renders the loaded page correctly', () => {

        const mockItems = createMockCharacters(10);

        const wrapper = shallow(
            <ItemMenu params={params} itemType="characters" />
        );
        wrapper.setState({ loading: false, totalPages: 10, items: mockItems });
        expect(toJson(wrapper)).toMatchSnapshot();
    });

});
