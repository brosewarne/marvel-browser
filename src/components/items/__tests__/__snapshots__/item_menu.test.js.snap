// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ItemMenu /> renders the loaded page correctly 1`] = `
<div>
  <section>
    <Component
      itemType="characters"
    />
  </section>
  <content>
    <WithStyles(Typography)
      variant="headline"
    >
      <strong>
        Browsing 
        A
         
        characters
         - Page 
        1
         of 
        10
      </strong>
    </WithStyles(Typography)>
    <Component
      itemTiles={
        Array [
          <Unknown
            item={
              Object {
                "id": 0,
                "name": "Character_0",
                "thumbnail": Object {
                  "extension": "png",
                  "path": "/a/href/to/an/image",
                },
              }
            }
            itemType="characters"
          />,
          <Unknown
            item={
              Object {
                "id": 1,
                "name": "Character_1",
                "thumbnail": Object {
                  "extension": "png",
                  "path": "/a/href/to/an/image",
                },
              }
            }
            itemType="characters"
          />,
          <Unknown
            item={
              Object {
                "id": 2,
                "name": "Character_2",
                "thumbnail": Object {
                  "extension": "png",
                  "path": "/a/href/to/an/image",
                },
              }
            }
            itemType="characters"
          />,
          <Unknown
            item={
              Object {
                "id": 3,
                "name": "Character_3",
                "thumbnail": Object {
                  "extension": "png",
                  "path": "/a/href/to/an/image",
                },
              }
            }
            itemType="characters"
          />,
          <Unknown
            item={
              Object {
                "id": 4,
                "name": "Character_4",
                "thumbnail": Object {
                  "extension": "png",
                  "path": "/a/href/to/an/image",
                },
              }
            }
            itemType="characters"
          />,
          <Unknown
            item={
              Object {
                "id": 5,
                "name": "Character_5",
                "thumbnail": Object {
                  "extension": "png",
                  "path": "/a/href/to/an/image",
                },
              }
            }
            itemType="characters"
          />,
          <Unknown
            item={
              Object {
                "id": 6,
                "name": "Character_6",
                "thumbnail": Object {
                  "extension": "png",
                  "path": "/a/href/to/an/image",
                },
              }
            }
            itemType="characters"
          />,
          <Unknown
            item={
              Object {
                "id": 7,
                "name": "Character_7",
                "thumbnail": Object {
                  "extension": "png",
                  "path": "/a/href/to/an/image",
                },
              }
            }
            itemType="characters"
          />,
          <Unknown
            item={
              Object {
                "id": 8,
                "name": "Character_8",
                "thumbnail": Object {
                  "extension": "png",
                  "path": "/a/href/to/an/image",
                },
              }
            }
            itemType="characters"
          />,
          <Unknown
            item={
              Object {
                "id": 9,
                "name": "Character_9",
                "thumbnail": Object {
                  "extension": "png",
                  "path": "/a/href/to/an/image",
                },
              }
            }
            itemType="characters"
          />,
        ]
      }
    />
    <Component
      loading={false}
      startsWith="A"
      totalPages={10}
    />
  </content>
</div>
`;

exports[`<ItemMenu /> renders the loading spinner correctly 1`] = `
<div>
  <section>
    <Component
      itemType="characters"
    />
  </section>
  <content>
    <div>
      <h1>
        LOADING 
        CHARACTERS
        .....
      </h1>
      <div>
        <img
          alt="loading"
          className="loadingSpinner"
          src="loading.svg"
        />
      </div>
    </div>
    <Component
      loading={true}
      startsWith="A"
      totalPages={0}
    />
  </content>
</div>
`;
